contracts/verifiers/UniversalVerifierMultiQuery.sol

```diff
@@ -0,0 +1,801 @@
+        }
+        // check that all the requests are set before adding them to the group
+        // because there is a check in setRequest that the group doesn't exist yet
+        for (uint256 i = 0; i < requestIds.length; i++) {
+            s._groupedRequests[groupId].push(requestIds[i]);
+        }
+    }
+
+    /**
+     * @dev Gets a specific request by ID
+     * @param requestId The ID of the request
+     * @return request The request data
+     */
+    function getRequest(
+        uint256 requestId
+    ) public view checkRequestExistence(requestId, true) returns (Request memory request) {
+        return _getUniversalVerifierMultiQueryStorage()._requests[requestId];
+    }
+
+    /**
+     * @dev Checks auth request in query
+     * @param query The query data
+     */
+    function _checkAuthRequestInQuery(Query calldata query) internal pure {
+        uint256 numAuthRequests = 0;
+
+        for (uint256 i = 0; i < query.requestIds.length; i++) {
+            if (_getRequestType(query.requestIds[i]) == AUTH_REQUEST_TYPE) {
+                numAuthRequests++;
+            }
+        }
+
+        require(numAuthRequests == 1, "Exactly 1 auth request is required");
+    }
+
+    /**
+     * @dev Sets a query
+     * @param queryId The ID of the query
+     * @param query The query data
+     */
+    function setQuery(
+        uint256 queryId,
+        Query calldata query
+    ) public checkQueryExistence(queryId, false) {
+        UniversalVerifierMultiQueryStorage storage s = _getUniversalVerifierMultiQueryStorage();
+
+        // check that exactly 1 auth request is included in the query
+        _checkAuthRequestInQuery(query);
+        s._queries[queryId] = query;
+        s._queryIds.push(queryId);
+
+        // check that all the requests of the groups are included in this query
+        _checkRequestsInGroupsIncluded(queryId);
+
+        emit QuerySet(queryId, query.requestIds);
+    }
+
+    /**
+     * @dev Gets a specific multi query by ID
+     * @param queryId The ID of the multi query
+     * @return query The query data
+     */
+    function getQuery(uint256 queryId) public view returns (Query memory query) {
+        return _getUniversalVerifierMultiQueryStorage()._queries[queryId];
+    }
+
+    /**
+     * @dev Checks for exactly 1 auth response and returns the index of the auth response
+     * @param responses The list of responses
+     * @return The index of the auth response
+     */
+    function _checkAuthResponses(Response[] memory responses) internal pure returns (uint256) {
+        uint256 numAuthResponses = 0;
+        uint256 authResponseIndex;
+
+        for (uint256 i = 0; i < responses.length; i++) {
+            if (_getRequestType(responses[i].requestId) == AUTH_REQUEST_TYPE) {
+                authResponseIndex = i;
+                numAuthResponses++;
+            }
+        }
+        require(numAuthResponses == 1, "Exactly 1 auth response is required");
```

Comment:
Technically we can not to provide auth response at all. It could be provided separately in other transaction.
I would simplify it in a way not to have any quantity restriction for the auth responses.